### ISSUE TITLE
Pruning on the basis of previous values of that trial and on the basis of values of previous values of other trials

### ISSUE THREAD

**Issue by schmnin** [optuna/optuna#2769]
I am currently trying to set up my first optuna project and struggle with the pruning mechanisms. I want to:
1. Stop the training, when the validation loss of my current trial increased for the last n (e.g. 5) epochs
2. Use pruning techniques like a median pruner to stop the trial if compared to other trials this trial does not perform well

I am using pytorch lighning and assume that the important part lies within the definition of the trainer in the objective function. At the moment, my trainer is defined as:

    trainer = pl.Trainer(
        max_epochs=10,
        progress_bar_refresh_rate=10,
        logger=logger,
        callbacks=[PyTorchLightningPruningCallback(trial, monitor='val_loss'),
                   EarlyStopping(monitor='val_loss', patience=5)],
        gpus=1 if torch.cuda.is_available() else None
    )

The pruner of my study is: 
pruner: optuna.pruners.BasePruner = (
    optuna.pruners.MedianPruner())

Before, I only used the MedianPruner which worked well. Now, my training curves look somewhat like this which obviously indicates that the pruning does not work the way I want it to work. 
![grafik](https://user-images.githubusercontent.com/86346375/123804638-af032600-d8ed-11eb-95ba-669ad0a71dea.png)

Looking forward to your suggestions and ideas :)

**Comment by nzw0301** [comment]
How about using [`PatientPruner`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.pruners.PatientPruner.html#optuna.pruners.PatientPruner) with MedianPruner?

**Comment by schmnin** [comment]
> 
> 
> How about using [`PatientPruner`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.pruners.PatientPruner.html#optuna.pruners.PatientPruner) with MedianPruner?

As far as I understand the documentation correctly this would only solve 2), since median pruner compares the monitored value with values from previous trials. 
What I ment with 1): I also want to compare the monitored value of the current epoch with the values of previous epochs, only considering the metric of this hyperparameter trial.

**Comment by nzw0301** [comment]
Ah, you're right! I'm not familiar with pytorch-lightning, so it might be wrong, but in my understanding, the plot looks strange. Could you share the minimal and reproducible code to generate it?

**Comment by schmnin** [comment]
Yes, thanks for taking the time to have a look at everything :) So this is the model and objective function definition:

```

class LSTMModel(pl.LightningModule):

    def __init__(self, lstm_dict, batch_size, n_features):
        """
        Initializes the network architecture

        Parameters:
          n_features (int): number of expected features, equals input size of LSTM
          hidden_size (int): number of hidden units in LSTM
          output_size (int): number of output units of output layer
          n_layers (int): number of LSTM layers
          do (float): Dropout for regularization
        """
        super().__init__()
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

        self.sequence_length = lstm_dict['sequence_length']
        #self.input_size = lstm_dict['n_features']
        self.input_size = n_features
        self.hidden_size = lstm_dict['hidden_size']
        self.output_size = lstm_dict['output_size']
        self.n_layers = lstm_dict['lstm_layers']
        self.do = lstm_dict['dropout_rate']
        self.lr = lstm_dict['lr']

        self.batch_size = batch_size

        self.hyperparams = lstm_dict

        self.lstm = nn.LSTM(input_size=self.input_size,
                            hidden_size=self.hidden_size,
                            num_layers=self.n_layers,
                            batch_first=True,  # input and output: (batch,seq,feature)
                            dropout=self.do).to(device)

        self.linear = nn.Linear(in_features=self.hidden_size,
                                out_features=self.output_size).to(device)

        self.training_losses = []
        self.validation_losses = []

    def reset_hidden_state(self, current_batch_size):
        """
        Initialize or reset the hidden state of LSTM

        Parameters:
          batch_size (int): batch size during training
        """
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        return torch.zeros(self.n_layers, current_batch_size, self.hidden_size).to(device)

    def forward(self, x):
        """
        Definition of forward flow

        Parameters:
          x (Tensor): input float tensor of shape (batch_size, seq_length, n_features)
        """
        current_batch_size = x.shape[0]
        hidden_state = self.reset_hidden_state(current_batch_size)
        cell_state = self.reset_hidden_state(current_batch_size)

        lstm_out, (h_out, _) = self.lstm(x, (hidden_state, cell_state))
        lstm_out = lstm_out[:, -1, :]
        out = self.linear(lstm_out)

        return out

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=self.lr)
        return optimizer

    def mse_loss(self, output, target):
        loss = F.mse_loss(output, target)
        return loss

    def training_step(self, train_batch, batch_idx):
        x, y = train_batch
        output = self.forward(x)
        loss = self.mse_loss(output, y)
        self.log('train_loss', loss)
        self.training_losses.append(loss)
        return loss

    def validation_step(self, val_batch, batch_idx):
        x, y = val_batch
        output = self.forward(x)
        loss = self.mse_loss(output, y)
        # self.log('val_loss', loss)
        self.validation_losses.append(loss)
        self.log('val_loss', loss)
        return loss

    def test_step(self, test_batch, batch_idx):
        x, y = test_batch
        output = self.forward(x)
        loss = self.mse_loss(output, y)
        self.log('test_loss', loss)

    def training_epoch_end(self, outputs) -> None:
        # avg_loss = torch.stack([x['loss'] for x in outputs]).mean()
        avg_loss = torch.mean(torch.stack(self.training_losses))
        self.training_losses = []
        self.logger.experiment.add_scalar("Loss/Train", avg_loss,
                                          self.current_epoch)

        if self.current_epoch == 1:
            # add model graph to logging
            device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
            sample = torch.rand((1, self.sequence_length, self.input_size)
                                ).to(device)  # (1, sequence_length, #number features)
            self.logger.experiment.add_graph(self, sample)
            # add hyperparameter info
            file_path = f"{self.trainer.logger.log_dir}/hparams.yaml"
            pl.core.saving.save_hparams_to_yaml(config_yaml=file_path, hparams=self.hyperparams)

    def validation_epoch_end(self, outputs) -> None:
        avg_loss = torch.mean(torch.stack(self.validation_losses))
        self.validation_losses = []
        self.logger.experiment.add_scalar("Loss/Validation", avg_loss,
                                          self.current_epoch)


def objective(trial):
    global model
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    lstm_dict = {'hidden_size': trial.suggest_int('hidden_size', 128, 512),
                 'output_size': divi_dict['target_length'],
                 'sequence_length': divi_dict['input_length'],
                 'lstm_layers': trial.suggest_int('lstm_layers', 1, 3),
                 'lr': trial.suggest_float("lr", 1e-5, 1e-2, log=True),
                 'dropout_rate': trial.suggest_float("dropout", 0.2, 0.5)}
    # define logger object
    logger = TensorBoardLogger(save_dir='tb_logs',
                               name='LSTM')
    trainer = pl.Trainer(
        max_epochs=25,
        progress_bar_refresh_rate=10,
        logger=logger,
        callbacks=[PyTorchLightningPruningCallback(trial, monitor='val_loss'),
                   EarlyStopping(monitor='val_loss', patience=5)],
        gpus=1 if torch.cuda.is_available() else None
    )
    dm = DIVIDataModule(data_path=PATH, batch_size=batch_size,
                        divi_dict=divi_dict, shuffle=True, test_setup=False)
    model = LSTMModel(lstm_dict=lstm_dict, batch_size=batch_size, n_features=n_features).to(device)
    trainer.logger.log_hyperparams(lstm_dict)
    trainer.fit(model, datamodule=dm)
    return trainer.callback_metrics["val_loss"].item()
```

they are used in the following lines of code: 
`
pruner: optuna.pruners.BasePruner = (
    optuna.pruners.MedianPruner())
study = optuna.create_study(direction="minimize",
                              pruner=pruner,
                              study_name='LSTM')
study.optimize(objective, n_trials=3,
               callbacks=[store_best_model])
`

Hope that that's everything you need :)

**Comment by nzw0301** [comment]
When I  just look at the codes, `MedianPruner` does not prune any trial since `n_trials=3` is less than `n_startup_trials` that is `5` by default. So maybe it does not relate to optuna...

**Comment by nzw0301** [comment]
Thanks! Unfortunately, the code is not reproducible, so I just looked at the codes. I think `MedianPruner` does not prune any trial since `n_trials=3` is less than `n_startup_trials` that is `5` by default. So I think the strange behaviour comes from pytorch-lightning's early-stopping, not optuna.

**Comment by schmnin** [comment]
Thanks for the idea, but I am pretty sure that that's not the problem: I normally save the study and continue it, so that in the end I'll have over 20 trials... Anyways, to get back to my original question: Should the combination of the pytorch early-stoppinng and optuna median pruner theoretically work? :)

**Comment by nzw0301** [comment]
yes, I think so.

**Comment by Crissman** [comment]
@schmnin Since the pruning issue you point out is being done by PyTorch early-stopping and not Optuna, is it ok to close this issue?

Thanks,
